In [1]:
import re
from test_utilities import run_tests_params

<link href="style.css" rel="stylesheet"></link>
<article class="day-desc"><h2>--- Day 3: Perfectly Spherical Houses in a Vacuum ---</h2><p>Santa is delivering presents to an infinite two-dimensional grid of houses.</p>
<p>He begins by delivering a present to the house at his starting location, and then an elf at the North Pole calls him via radio and tells him where to move next.  Moves are always exactly one house to the north (<code>^</code>), south (<code>v</code>), east (<code>&gt;</code>), or west (<code>&lt;</code>).  After each move, he delivers another present to the house at his new location.</p>
<p>However, the elf back at the north pole has had a little too much eggnog, and so his directions are a little off, and Santa ends up visiting some houses more than once.  How many houses receive <em>at least one present</em>?</p>
<p>For example:</p>
<ul>
<li><code>&gt;</code> delivers presents to <code>2</code> houses: one at the starting location, and one to the east.</li>
<li><code>^&gt;v&lt;</code> delivers presents to <code>4</code> houses in a square, including twice to the house at his starting/ending location.</li>
<li><code>^v^v^v^v^v</code> delivers a bunch of presents to some very lucky children at only <code>2</code> houses.</li>
</ul>
</article>


In [4]:
tests = [
    {
        "name": "Example 1",
        "ds": ">",
        "expected": 2,
    },
    {
        "name": "Example 1",
        "ds": "^>v<",
        "expected": 4,
    },
    {
        "name": "Example 1",
        "ds": "^v^v^v^v^v",
        "expected": 2,
    },
]


def houses(ds: str) -> int:
    houses = {(0, 0)}
    r, c = 0, 0
    for d in ds:
        if d == ">":
            c += 1
        elif d == "<":
            c -= 1
        elif d == "^":
            r -= 1
        else:  # d == 'v':
            r += 1
        houses.add((r, c))

    return len(houses)


run_tests_params(houses, tests)


Test Example 1 passed, for houses.
Test Example 1 passed, for houses.
Test Example 1 passed, for houses.
Success


In [5]:
with open("../input/day3.txt") as f:
    print(houses(f.read()))

2565


<link href="style.css" rel="stylesheet"></link>
<main>

<p>Your puzzle answer was <code>2565</code>.</p><p class="day-success">The first half of this puzzle is complete! It provides one gold star: *</p>
<article class="day-desc"><h2 id="part2">--- Part Two ---</h2><p>The next year, to speed up the process, Santa creates a robot version of himself, <em>Robo-Santa</em>, to deliver presents with him.</p>
<p>Santa and Robo-Santa start at the same location (delivering two presents to the same starting house), then take turns moving based on instructions from the elf, who is <span title="This absolutely real word was invented by someone flipping eggnoggedly through a dictionary.">eggnoggedly</span> reading from the same script as the previous year.</p>
<p>This year, how many houses receive <em>at least one present</em>?</p>
<p>For example:</p>
<ul>
<li><code>^v</code> delivers presents to <code>3</code> houses, because Santa goes north, and then Robo-Santa goes south.</li>
<li><code>^&gt;v&lt;</code> now delivers presents to <code>3</code> houses, and Santa and Robo-Santa end up back where they started.</li>
<li><code>^v^v^v^v^v</code> now delivers presents to <code>11</code> houses, with Santa going one direction and Robo-Santa going the other.</li>
</ul>
</article>

</main>


In [10]:
tests = [
    {
        "name": "Example 1",
        "ds": "^v",
        "expected": 3,
    },
    {
        "name": "Example 1",
        "ds": "^>v<",
        "expected": 3,
    },
    {
        "name": "Example 1",
        "ds": "^v^v^v^v^v",
        "expected": 11,
    },
]


def houses_with_robo(ds: str) -> int:
    houses = {(0, 0)}
    r_s, c_s = 0, 0
    r_r, c_r = 0, 0
    santa = True
    for d in ds:
        r, c = (r_s, c_s) if santa else (r_r, c_r)
        if d == ">":
            c += 1
        elif d == "<":
            c -= 1
        elif d == "^":
            r -= 1
        else:  # d == 'v':
            r += 1
        houses.add((r, c))
        if santa:
            r_s, c_s = r, c
        else:
            r_r, c_r = r, c

        santa = not santa
    return len(houses)


run_tests_params(houses_with_robo, tests)


Test Example 1 passed, for houses_with_robo.
Test Example 1 passed, for houses_with_robo.
Test Example 1 passed, for houses_with_robo.
Success


In [11]:
with open("../input/day3.txt") as f:
    print(houses_with_robo(f.read()))

2639


<link href="style.css" rel="stylesheet"></link>

<main>

<p>Your puzzle answer was <code>2639</code>.</p><p class="day-success">Both parts of this puzzle are complete! They provide two gold stars: **</p>
<p>At this point, you should <a href="/2015">return to your Advent calendar</a> and try another puzzle.</p>
<p>If you still want to see it, you can <a href="3/input" target="_blank">get your puzzle input</a>.</p>
<p>You can also <span class="share">[Share<span class="share-content">on
  <a href="https://twitter.com/intent/tweet?text=I%27ve+completed+%22Perfectly+Spherical+Houses+in+a+Vacuum%22+%2D+Day+3+%2D+Advent+of+Code+2015&amp;url=https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F3&amp;related=ericwastl&amp;hashtags=AdventOfCode" target="_blank">Twitter</a>
  <a href="javascript:void(0);" onclick="var ms; try{ms=localStorage.getItem('mastodon.server')}finally{} if(typeof ms!=='string')ms=''; ms=prompt('Mastodon Server?',ms); if(typeof ms==='string' &amp;&amp; ms.length){this.href='https://'+ms+'/share?text=I%27ve+completed+%22Perfectly+Spherical+Houses+in+a+Vacuum%22+%2D+Day+3+%2D+Advent+of+Code+2015+%23AdventOfCode+https%3A%2F%2Fadventofcode%2Ecom%2F2015%2Fday%2F3';try{localStorage.setItem('mastodon.server',ms);}finally{}}else{return false;}" target="_blank">Mastodon</a></span>]</span> this puzzle.</p>
</main>
